autoprecompiles/src/memory_optimizer.rs

```diff
@@ -0,0 +1,371 @@
+    let mut new_constraints: Vec<SymbolicConstraint<T>> = Vec::new();
+
+    // Go through the memory interactions and try to optimize them.
+    let mut memory_contents: HashMap<_, (Option<usize>, Vec<AlgebraicExpression<_>>)> =
+        Default::default();
+    let mut to_remove: Vec<usize> = Default::default();
+
+    for (index, bus_int) in machine.bus_interactions.iter().enumerate() {
+        let mem_int = match MemoryBusInteraction::try_from_symbolic_bus_interaction_with_memory_kind(
+            bus_int,
+            MemoryType::Memory,
+        ) {
+            Ok(Some(mem_int)) => mem_int,
+            Ok(None) => continue,
+            Err(_) => {
+                // This interaction might be going to memory, but we do not know
+                // the multiplicity. Delete all knowledge.
+                memory_contents.clear();
+                continue;
+            }
+        };
+        let addr = algebraic_to_quadratic_symbolic_expression(&mem_int.addr);
+
+        match mem_int.op {
+            MemoryOp::Receive => {
+                if let Some((previous_send, existing_values)) = memory_contents.get(&addr) {
+                    for (existing, new) in existing_values.iter().zip(mem_int.data.iter()) {
+                        if existing != new {
+                            new_constraints.push(
+                                AlgebraicExpression::new_binary(
+                                    existing.clone(),
+                                    AlgebraicBinaryOperator::Sub,
+                                    new.clone(),
+                                )
+                                .into(),
+                            );
+                        }
+                    }
+
+                    // If we got this information from a previous send, we can remove both.
+                    if let Some(previous_send) = previous_send {
+                        to_remove.extend([index, *previous_send]);
+                    }
+                } else {
+                    memory_contents.insert(addr.clone(), (None, mem_int.data.clone()));
+                }
+            }
+            MemoryOp::Send => {
+                // We can only retain knowledge about addresses where we can prove
+                // that this send operation does not interfere with it, i.e.
+                // if we can prove that the two addresses differ by at least a word size.
+                memory_contents.retain(|other_addr, _| {
+                    address_comparator.are_addrs_known_to_be_different_by_word(&addr, other_addr)
+                });
+                memory_contents.insert(addr.clone(), (Some(index), mem_int.data.clone()));
+            }
+        }
+    }
+
+    log::debug!(
+        "Removing {} memory interactions and adding {} new constraints",
+        to_remove.len(),
+        new_constraints.len()
+    );
+
+    (to_remove, new_constraints)
+}
+
+trait AddressComparator<A> {
```

Comment:
The trait is currently not actually used, but I will implement a simpler one for the register optimizer and then we can use a single tool for both.